### PR TITLE
Bitfield improvements

### PR DESCRIPTION
**Pull request by rsmmr**
This addresses #1032, #1468,  and #1484.

### CODE REVIEW

File: spicy/toolchain/bin/spicy-dump/printer-text.cc [hunk 1]
```diff
@@ -116,7 +136,11 @@ void TextPrinter::print(const type_info::Value& v) {
                     if ( y ) {
                         out() << '\n';
                         outputIndent();
-                        out() << f.name << ": ";
+
+                        if ( ! f.isAnonymous() )
+                            out() << f.name;
```
Comment (bbannier) [inline spicy/toolchain/bin/spicy-dump/printer-text.cc:141]: I believe this is still inconsistent and not what we want. It seems that e.g., `spicy-dump` omits any anonymous field unless it is a bitfield, e.g., given the following unit
```ruby
public type X = unit {
    : uint8;
    : bitfield(8) { a: 0..7; };
};
```
I see the following output
```console
$ printf '\x00\x00' | spicy-dump bla.spicy
bla::X {
  : (a: 0)
}
```
I would have expected no output here.

**Review by bbannier — CHANGES_REQUESTED** [review]

File: spicy/toolchain/src/compiler/visitors/normalizer.cc [hunk 2]
```diff
@@ -227,6 +227,48 @@ struct Visitor : public hilti::visitor::PostOrder<void, Visitor> {
...
+                auto bits = t->bits(id);
+                if ( ! bits )
+                    continue;
+
+                if ( found_field ) {
+                    hilti::logger().error(
+                        hilti::util::fmt("unit '%s' has multiple anonymous bitfields with member '%s'", *unit->id(),
+                                         id));
+                    return;
+                }
```
[the remark this answers is not in the document]
Comment (rsmmr) [reply spicy/toolchain/src/compiler/visitors/normalizer.cc:255]: > Just wondering whether we should instead reject multiple anon bitfields with fields with same name

Ack, I'll add that validation.

File: spicy/toolchain/bin/spicy-dump/printer-text.cc (hunk 1 above)
Comment (rsmmr) [reply spicy/toolchain/bin/spicy-dump/printer-text.cc:141]: Agree that that's not right. However, in the spirit of our use case here, what I'd expect is actually this:

```
bla::X {
  a: 0
}
```

I'll give that a try.

Comment (rsmmr) [reply spicy/toolchain/bin/spicy-dump/printer-text.cc:141]: Adding three new commits at the end implementation this. The first two change how (normal) bitfields and also offsets are printed, the 3rd then implements printing of anonymous bitfields.

**Review by bbannier — APPROVED** [review]
🚀

**Pull request merged**